telemetry/convert.py

```diff
@@ -146,10 +149,127 @@ def convert_json(self, jsonstr, date, ip=None):
+        payload_info = payload["info"]
+        for field in payload_info:
+            if field not in json_dict["info"]:
+                json_dict["info"][field] = payload_info[field]
+        # Back up payload.ver
+        json_dict["payload.ver"] = payload["ver"]
+
+        # Get rid of duplicated data in the payload field
+        del json_dict["payload"]
+
+        # Step 2 of conversion:
+        #   Recreate the old-style ping.info section from fields that are
+        #   now in ping.environment
+        envFields = json_dict["environment"]
+        info = json_dict["info"]
+        self.add_info_fields(info, envFields, envFieldMap)
+
+        adapters = None
+        try:
+            adapters = envFields["system"]["gfx"]["adapters"]
+        except KeyError, TypeError:
+            pass
+
+        if type(adapters) == list and len(adapters):
+            self.add_info_fields(info, adapters[0], adapterFieldMap)
+            if len(adapters) > 1:
+                # Copy details of the second GPU
+                self.add_info_fields(info, adapters[1], adapterFieldMap, "2")
+                info["isGPU2Active"] = bool(adapters[1].get("GPUActive"))
+
+        # Fix up clientID
+        json_dict["clientID"] = json_dict.get("clientId")
+
+        # Step 3:
+        #   Add a "ver" field to make it look like a classic ping, and
+        #   delete the "version" field to make the conversion idempotent.
+        #   This ping will then be passed through convert_obj again to
+        #   fix up the histograms.
+        json_dict["original.version"] = json_dict["version"]
+        del json_dict["version"]
+        json_dict["ver"] = Converter.VERSION_UNCONVERTED
+
     def convert_obj(self, json_dict, date, ip=None):
-        info = json_dict.get("info", None)
-        # Check if the payload is already converted:
-        if "ver" in json_dict:
+        if "version" in json_dict:
+            # This is a new unified (FHR+Telemetry) ping
+            if json_dict["version"] != Converter.VERSION_UNIFIED:
+                raise ValueError("Unknown unified ping version: " + str(json_dict["version"]))
+            elif "type" not in json_dict:
+                raise ValueError("Unified ping has no top-level 'type' field")
+            elif "application" not in json_dict:
+                raise ValueError("Unified ping has no top-level 'application' field")
+            elif "payload" not in json_dict:
+                raise ValueError("Unified ping has no top-level 'payload' field")
+
+            # Create a ping.info field and copy select ping.application fields.
+            # This is the minimum required for compatibility with the calling
+            # process_incoming* scripts
+            info = json_dict["info"] = {}
+            self.add_info_fields(info, json_dict["application"], appFieldMap)
+            if json_dict["payload"].get("info"):
+                info["reason"] = json_dict["payload"]["info"].get("reason")
```

Comment:
Hmm, actually you want to force the reason to "saved-session" to ensure compatilibility with subsequent analyses, right?